Native reference guide updates (#24013)
quarkusio/quarkus: docs/src/main/asciidoc/native-reference.adoc

it needs to be escaped to be processed correcly, e.g. `\\,`.
====

== Inspecting Native Executables

Given a native executable, various Linux tools can be used to inspect it.

> To allow support a variety of environments,

Reviewer: ```suggestion
To allow supporting a variety of environments,
```

inspections will be done from within a Linux container.

> Let's create a Linux container image that with all the tools required for this guide:

Reviewer: ```suggestion
Let's create a Linux container image with all the tools required for this guide:
```

Reviewer 2: Isn't something missing after the `that`?

Author: Good point, I don't remember anything missing that I wanted to say. I was probably simply typing one way to say it and ended up with the other :)

This is useful for getting an approximation on why a method/class is included in the binary.
However, the text format makes it very difficult to read and can take up a lot of space.

> Since Mandrel 21.3.0.0, the call tree is also reported as a group of CSV files.
> The CSV output can be enabled by adding `-H:PrintAnalysisCallTreeType=CSV` to the additional native arguments. E.g.

Reviewer 3: ```suggestion
Since Mandrel 22.0.0.0, the call tree is also reported as a group of CSV files.
The CSV output can be enabled by adding `-H:PrintAnalysisCallTreeType=CSV` to the additional native arguments. E.g.
```

`-H:PrintAnalysisCallTreeType` was introduced in 22.0.0.0, it's not available in 21.3.0.0. In 21.3.0.0 `-Dquarkus.native.enable-reports` (i.e. `-H:PrintAnalysisCallTree`) will produce both the text-formatted call tree and the CSV files.

We might need to consider sticking to 21.3.0.0 in the guide since it's an LTS

Author: The updates in this PR are for `main`, which is already past the LTS (2.7 / 21.3) and it's already using Mandrel 22.0.

Author: I might send a PR specific for 2.7 branch (the LTS) adjusting things as you suggest above since that's on 21.3.

[source,bash,subs=attributes+]
----
./mvnw package -DskipTests -Pnative \

>     -Dquarkus.native.enable-reports \

Reviewer 3: ```suggestion
```
`-Dquarkus.native.enable-reports` (i.e. `-H:PrintAnalysisCallTree`) is implied by `-H:PrintAnalysisCallTreeType=CSV`.

Author: Hmmm, but the user might not be aware of this subtlety. We can't expect them to be aware of the secondary effect that advanced options have. I'd rather be explicit here and keep `enable-reports`.

Side note, the need for the extra option will eventually go away https://github.com/quarkusio/quarkus/issues/20733. I'll tackle that next few days now that `main` is on 22.0.

Copy and paste the contents of the script into a file called `import.cypher`.

[WARNING]
====

> Mandrel 22.0.0 contains a bug where the symbolic links used by the import cypher file are not correctly set when generating reports within a container

Reviewer 3: This looks like a good reason to use 21.3.x.x instead.

Author: Hmmm, it's a very small issue that can easily be worked around. I don't agree.

docker exec testneo4j bin/cypher-shell -u neo4j -p ${NEO_PASS} -f import.cypher
----

[TIP]
====

> If the Neo4j import fails, you might need to delete leftover data before importing it again.

Reviewer 3: killing and rerunning the `testneo4j` container might be simpler/safer.

Author: Ah yes, good point 😅

Since we're dealing with a linux native executable,
we can use tools like `perf` directly.
To use `perf`,
start the tools container created earlier as a privileged user:

[source,bash]
----
cd debugging-native

> docker run --privileged -t -i --rm -v ${PWD}:/data -p 8080:8080 fedora-tools:v1 /bin/bash

Reviewer 3: ```suggestion
docker run --privileged -t -i --rm -v ${PWD}:/data -p 8080:8080 fedora-tools:v1
```

----

[[NOTE]]
====
Note that in order to use `perf` to profile the native executables in the guide,
the container needs to run as privileged, or with `--cap-add sys_admin`.
Please note that privileged containers are **NOT** recommended in production, so use this flag with caution!
====

Once the container is running, you need to ensure that the kernel is ready for the profiling exercises:

[source,bash]
----
echo -1 | sudo tee /proc/sys/kernel/perf_event_paranoid
echo 0 | sudo tee /proc/sys/kernel/kptr_restrict
----

[TIP]
====

> The kernel modifications above also applied to Linux virtual machines.

Reviewer 3: ```suggestion
The kernel modifications above also apply to Linux virtual machines.
```

If running on a bare metal Linux machine,

> tweaking `perf_event_paranoid` is enough.

Reviewer 3: ```suggestion
tweaking only `perf_event_paranoid` is enough.
```

====

Then, from inside the tools container we execute:

[source,bash]
----
perf record -F 1009 -g -a ./target/debugging-native-1.0.0-SNAPSHOT-runner
----

While `perf record` is running, open another window and access the endpoint:

[source,bash]
----
curl http://localhost:8080/string-builder # this will never complete
----

After a few seconds, halt the `perf record` process.
This will generate a `perf.data` file.
We could use `perf report` to inspect the perf data,
but you can often get a better picture showing that data as a flame graph.
To generate flame graphs, we will use
https://github.com/brendangregg/FlameGraph[FlameGraph GitHub repository],

> which it has already been installed inside the tools container.

Reviewer: ```suggestion
which has already been installed inside the tools container.
```
